### PR TITLE
scripts: gen_defines.py: Add out_dev() flag for deprecated macros

### PR DESCRIPTION
**Pull request by ulfalizer**
Add a 'deprecation_msg' string/flag to out_dev(). When 'deprecation_msg'
is passed, all generated macros include

    __WARN(<deprecation_msg>)

which prints a custom warning if the macro is used.

Meant to be used when improving the output format.

### CODE REVIEW

**Comment by ulfalizer** [comment]
This commit also appears in https://github.com/zephyrproject-rtos/zephyr/pull/21506. Adding it separately in case that PR gets stalled.

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -771,6 +776,17 @@ def out(ident, val, aliases=()):
     return primary_ident
 
 
+def out_define(ident, val, deprecation_msg, out_file):
+    # out() helper for writing a #define. See out_dev() for the meaning of
+    # 'deprecation_msg'.
+
+    s = "#define DT_{:40}".format(ident)
+    if deprecation_msg:
+        s += r' _Pragma("GCC warning \"{}\"")'.format(deprecation_msg)
```
Comment (galak) [inline scripts/dts/gen_defines.py]: Can we have this be __DEPRECATED_MACRO instead.  So that way this isn't GCC specific.

**Review by galak — CHANGES_REQUESTED** [review]
Make what we produce in the define use `__DEPRECATED_MACRO` so its not GCC specific.

**Comment by galak** [comment]
looks like we need to introduce a `__DEPRECATED_MACRO_MSG` that allows passing a message.

**Comment by ulfalizer** [comment]
Rebased now. Depends on https://github.com/zephyrproject-rtos/zephyr/pull/22336, so marking DNM until that one is in.

**Comment by galak** [comment]
Removing DNM now that #22336 is merged.

**Review by galak — APPROVED** [review]

**Pull request merged**